adhocteam/Head-Start-TTADP
# Add frontend user login and logout

**Description of change**

A user that has not logged in will be shown a 'landing' page with a login button presented. Once logged in the user will view a home page with a logout button, and the header will populate with links (home and the admin page are the only links right now). When the logout button is pressed a logout message is displayed.

The API will now return a 401 when accessed without a valid logged in session. Hitting `http://localhost:8080/api/login` (`http://localhost:3000/api/login` when proxied) will trigger the login process. 

**How to test**

 1) Populate `LOGIN_REDIRECT_URL` in the server's `.env` file (example and value you should use is in `.env.example`)
 2) `yarn deps` (`yarn docker:deps`)
 3) `yarn start` (`yarn docker:start`)
 4) Browse to `http://localhost:3000`. You should see a page prompting you to login with no navigation links
 5) Login to HSES via the button on step 3
 6) After login you should be back at our frontend and see "Welcome to the TTA Smart Hub" followed by your email address
 7) Click "logout" and you will be returned to the page on step 3

**Notes**

 * ~@SarahJaine we will have a new env variable to set for the backend. It is not a secret. It's name is `LOGIN_REDIRECT_URL` and is the url the server should redirect to after login to HSES (so the url of the frontend). I can poke around and figure out where to add it just a heads up that I might ping you if I can't figure it out.~ Figured it out, see comment below
 * ~An env variable was added to the frontend (named `REACT_APP_API_URL`) as well that is not a secret (we can't have frontend secrets anyways because they will always be sent along with the frontend HTML/javascript in plaintext). It will need to be set to the backend API URL in each environment at build time. @SarahJaine I might ping you again if I can't figure it out.~
 * API routes are now prefixed with `/api`. The OAuth callback hasn't been updated (we can't directly do the update) so it looks a bit goofy still having a non-prefixed route. It was required to keep login working. If we decide we want to prefix with `/api` and the callback has been updated in HSES it will be an easy switch for the backend. 
 * I imagine there will be updates to the user endpoint and calling code in the frontend once we create the users/permissions database tables

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/4

**Checklist**

- [n/a] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
